### ISSUE TITLE
Ability to create user libraries of STACK code for complex question families

### ISSUE THREAD

**Issue by sgparry** [maths/moodle-qtype_stack#474]
Currently STACK has very limited support for strictly 'core' libraries, that are loaded globally for a very short whitelist. Extending that set requires developer intervention. This is understandable as the libraries at this level have privileged access to the whole system.
For complex question families (e.g. base n), it would be really useful to be able to create and load STACK code from a common user repository. Code here would stored in the database as shared resource and exclusively edited via the STACK UI and hence could be subject to all the normal STACK restrictions. The question author would then include the 'library' code in her question by entering an include directive in her question code. This would be replaced by the PHP with the library code before parsing it and checking it in the normal way. This allows the library code to do everything normal STACK code can do, but no more - you in no way increase the attack surface.

**Comment by aharjula** [comment]
Your point of the attack surface and way of tying in the normal STACK security model into this is refreshingly sensible idea for dealing with the common problem of sharing complex code between questions. However, I will have to point out some problems, that will also need to dealt with:
 1. Version management and ownership of code. If there are multiple competent STACK authors in your VLE and one decides to change that library code how should the questions authored by others react? Will they just start using the new library code even though it might no longer work with them, or is there a reference to the old version and they keep using it, if so what if you want to update them to use the new one, all of them? If you update library code can you somehow test that everything referencing it still works or can be cleanly updated to reference the new version? 
 2. Export/import/backup, naturally, exported questions will also include the library code they need (as reference and the reference target, not as injected content so that one can restore this better situation), but when one imports such a question where does that code go, i.e., will it overwrite existing in the library if such exists and with what logic. What are the name spaces here? Will we try to find that same library code from our library and tie to that. What to do with export of multiple questions, or import of old versions of library code? What if two imports bring simillarilly versioned and named library code that differs from the other, will we rename the other library, which and how?

So basically, how would this library code be versioned and how will the access rights work, surely someone owns the fragments and can thus block others from modifying? Is this a feature for just the admin user of the VLE or can everyone with question editor access work with this.

```
/* So are we going for something like this? If so "Bob's lib" probably 
   won't be globally unique and will need work during import export.
   Also authors will not be writing that version number there, tools 
   will have to handle that */
include_shared("Bob's lib", "20200501-1412");
```

**Comment by sgparry** [comment]
All good points. My initial thoughts are: hash the library code on first entry and use that as the id number - this avoids at least 500 copies of the same code in the library. If and when the code is subsequently edited under the same name, or deleted, keep the old version with a "breadcrumb" linking to the new. At that point, ask the author: A) treat as new code B) voluntary upgrade everyone C) compulsory upgrade all my questions D) *danger* compulsory upgrade everyone's questions. This will then be recorded with the breadcrumb in the old version. In the question, store a list of hashes retrieved for the includes in the question. On the first run, just uses the latest version and store it. On subsequent edits, check for a later version. If one exists with a compulsory upgrade, do it and store it, if not ask. If just using / pewviewing the question assume no upgrade unless compulsory.

**Comment by sangwinc** [comment]
I'm certainly willing to discuss this further.  I am very worried about duplicate code between questions, but I also want to make sure questions are self-contained for import/export.

Any solution must work for STACK in both Moodle, ILIAS and other users (e.g. direct embedding via an API).  

I'd much rather receive small libraries from colleagues and add these to the core.  

To avoid delays in upgrade cycles, we could add in a "Maxima library" option to allow standard Moodle admins to include maxima code which would be available to any question.  That way advanced users can get on with developing libraries and individual questions without a developer cycle delay.

**Comment by aharjula** [comment]
Chris, I think it would not be a good idea to have a single "Maxima library" option at the system level, as that would mean that the code would be loaded into every question even if not needed. Also, a single common block would mean that if the authors have multiple differing "libraries" they need to be distinct enough that they do not affect each other. I also doubt whether all the things that one would handle with common code that would make sense to be included in the core, I would imagine that the code can simply be connected to a question-family and only make sense for that family of questions, it could also be that the code is used to declare/override the texput rules for a specific lecturers course.

I'll use the term fragment instead of library, as I do not see that all of these are fully finalised libraries with multiple functions and documentation. I see these as common code segments that are by themselves "inert" but can be glued into questions to augment the toolset available to the author.

In my opinion, the features described in this issue in 2019 are still sensible, with one difference, the inclusion should not be something written within the question-variables it needs to happen as a separate model that has database-integrity checks. The difficult bit is to provide question-editor level UI where one selects which library-fragments to load into the question and has the option of adding new or updating old fragments (with all the upgrade logic for dealing with ownership and versioning). Import/export are not really problems for this, if you export a question it brings the fragments with it. If I would design a data-model for this I would add two tables to the database:
```
..stack_fragments:
 - id, everything must have a DB-row-id
 - hash, the fragment content hash, for versioning like Stephen described, alternatively we could simply hold the id of the previous version of this fragment and create a new fragment every time modifications happen
 - timestamp, when this fragment was last modified, mostly for UI listings like the label
 - creator, the user that created this
 - code, the code of the fragment
 - label, a short label/name for this fragment, to be shown where ever fragments connected to questions are listed
 - description, longer text about the features of the fragment for when you are searching for a fragment that has specific features
 - compiled, 4.3 secure_loader style cache, if we pile these into the question we do not want to validate them every time, especially as these are probably longer bits of code...

...question_used_fragments:
 - id, everything must have a DB-row-id
 - questionid, the question using a fragment
 - fragmentid, the fragment used
 - order, integer if the question loads multiple fragments it is (it truly is) necessary to define the order of loading
```
Chris, you should also note that the `...stack_fragments` table has absolutely no need for questions to exist, and we could easily provide starter fragments in STACK-releases without any need for the releases to load in questions, the DB-upgrade script could simply load them in when new are added.

The difficult part is just the UI I see that at least the following UI features are needed:
 1. The question-editor listing of the fragments connected to the question. With remove and ordering.
 2. The question-editor tool for adding a fragment to the question, this needs to be able to provide a search of the fragment library. This needs to be able to also edit those fragments and trigger the whole updates to other questions logic.
 3. System-level access to the library, for selectively cleaning out old no longer used fragments (except those coming with STACK). Also linking to the question using those fragments.

**Comment by sangwinc** [comment]
Ok Matti, that makes sense.  Quite a lot of potential complexity....

**Comment by aharjula** [comment]
Chris, the versioning and ownership issues, when updating the fragments, are the biggest problem here just picking content to prepend to the question-variables is trivial. In any case, this needs to be something the question asks for not something that applies to the whole system while I mostly worry about the performance implications of the whole system approach what scares me more is the possibility of identifiers clashing in the common-code and question-code for questions that did not want that common-code.

While writing that last wall of text I really started to see the benefit that "starter fragments" could bring, it would at the same time provide multiple things:
 1. It would invite the authors to use this feature and thus improve the world.
 2. It would allow us to provide well tested and developed logic fragments, which have passed multiple eyeballs, from which to start, e.g. providing complex examples for some common texput magic etc. Best of all those fragments would be documented.
 3. It could be one way of implementing future unit-system configuration ideas, i.e. load a fragment that throws the SI system away and replaces it with the imperial one or some merger of them. Although, if the fragments start to affect the questions state/configuration by changing the defined units the CASSecurity objects will need to be notified.

**Comment by sangwinc** [comment]
That is true Matti,
But, that would still require authors to share this stuff with us!
Chris

**Comment by aharjula** [comment]
Share what? The idea is that the (sets of) questions come with their fragments and the authors produce them. The "starter fragment" concept would just be some initial fragments that we could provide. Maybe those could be in part something that authors share with others through us, but that would not be required.

As the fragments travel with questions there is no need for us to know about them so no one needs to share anything.

**Comment by aharjula** [comment]
In short, I would compare that `..stack_fragments` table to a bookshelf with a unique-filter in the local system. Every question may carry with it some additional books, and if they are not already present in that bookshelf they will be added to it when the question is imported. When a question is being authored that author may browse the shelf and select some books from there to be included in the question, or they may write a new one and add it to the collection. The "starter fragment" concept would simply place some default books to the self. The shelf can also contain books no one ever reads...

**Comment by sangwinc** [comment]
Ok Matti, I see the model.

If users edit these fragments, will the system automatically run every test on every version of every question which uses the updated fragment?  We could easily and unwittingly change something and break a whole load of questions.... 

A classic version control issue.

**Comment by aharjula** [comment]
I would go for COW here and leave the other questions referencing the old version by default. Provide the logic to query the author about which references to update in the way Stephen described, essentially this was already pretty much covered on the concept level last year. The key thing to note is the fact that one can only mess with questions one has the rights to edit. Easiest to consider the fragments as WORM objects that include a linked tree of historical versions so that updates can be done when a newer version is detected to be present. 

Basically, when one opens a question editor we would check if the question references fragments and if there are newer versions of those fragments asks if one wants to update to them, maybe gives a preview and tests with the new fragments, the preview and test pages could have some option that would allow overriding the fragments with newer ones for the purpose of testing.

**Comment by bkonev** [comment]
If I may add a suggestion: what if the stack fragments are kept at an external source (GitHub) and get cached locally? 
This could solve the export question as they could be downloaded on import.

In the first instance, the workflow could be quite straightforward: 
1. There should be optional fields for a GitHub repo and a file within that repo. No user identification in the first instance, downloads are only possible from public repos.
2. When a question is being saved, one could compute an md5 hash of the repo+file string and check if the directory with such a name already exists in a "packages" directory and if the file with the given name exists in that directory.
2a. Assuming both are OK, the content of the file can be simply prepended to the question variables. It would be nice to have the "show content" and "check for updates" buttons. 
2b. If the directory / file don't exist, one could attempt to download the package from GitHub. 
With the use of specific revisions/tags, one could address the issue of incompatible changes.

The main advantage is that this will encourage users to share their libraries, libraries will get tested and improved. One could create space where users would be able to advertise their libraries. Popular packages could then be absorbed into STACK itself.

**Comment by aharjula** [comment]
Interesting idea and does indeed have those benefits that you mentioned. Probably authors would not want to write revision hashes into their references so tags would sound better but naturally requires that tags exist in the repository.

Would be interesting to see someone taking this forward, the core developers do probably agree that this whole "fragments" thing is something that would be nice to have but none have had the need and the time to explore this. I do believe that this might be simpler after certain core features have first been changed, in the future we will probably have logic for caching the validation of the question-variables and plugging in any fragments directly to that cached content would be the logical place for doing this logic as it could take a bit longer there than in the current form of doing this every time the question-vars are used. No matter what DB-changes are inevitable as we need to list the referenced fragments somewhere.

Assuming Git is that only external source that would be supported it would also nicely externalise the version number tracking, i.e. one could check for potential upgrades through it, though reading the git log/tags from PHP might be somewhat complex.

**Comment by mkraska** [comment]
There has been no comment here since the release of STACK 4.4 with the include options for question variables and question text. I just want to give a short report on my observations.

In the last months I added such includes to nearly all of my appox. 300 questions, mainly to enrich numerical and algebraic input fields with formative standard feedback, see function [fb_unit()](https://extmoodle.th-brandenburg.de/mod/quiz/view.php?id=6065) (link is to a demo moodle course with option for guest login). This seems to be a good candidate of a new or extended standard answer test. It has undergone many edits due to feedback from bulk production and usage in classroom. 

Many of the questions also have JSXGraph widgets (mostly static, some interactive). These widgets make use of a standardized way to control them by maxima lists (meclib concept, see [a contribution to STACK22](https://doi.org/10.3991/ijet.v17i23.36501).

So STACK 4.4 gave a great boost to feature enrichment and consistency of my teaching material.

Yet the drawback is a severe slowdown in editing. A single debugging cycle of editing, saving, preview takes more than half a minute to perform. In particular, the meclib graphics frequently require multiple iterations for adjusting labels and testing for robustness of randomization.

Fortunately, this delay applies only once, so the students only suffer, when they hit a random version as the first participant.
Pre-emptive deployment of random versions also takes nearly a minute per variant, so doing this in an interactive editing session is painful. Yet I am not sure that this saves the students from accidentally hitting a non-cached/compiled version. 

Once a question is cached/linked, the external link targets aren't accessed any more, rather the internal result of compilation is used when the question is invoked by the users.

My include statments always link to the current version of the files, i.e. not to a static release. Updates of the targets don't have effects to existing compiled questions. They get upgraded only after some (perhaps dummy) code edit of the question.

The whole process is somewhat intransparent. There is no way to see the compilation status (including a time stamp) in the question bank, nor can you start bulk compilation as a batch job, at least not on course level (I am not a moodle admin).

**Comment by sangwinc** [comment]
Thanks for suggesting this.  I can't login as a guest and see the question!  
Can you point me to your include code?  I'm happy to review it and talk about getting this into the STACK core if possible.
Chris

**Comment by mkraska** [comment]
It seems that guests can't access tests by default. So I added this right to guest for the given question. Access as guest should be possible by now. Thank you for pointing out this issue.

To access all questions, you would need to create a login on our external server. 

Other references:

[Wiki entry](https://github.com/mkraska/meclib/wiki/fb_unit)

[Source file](https://github.com/mkraska/meclib/blob/main/Maxima/fb_value.mac)

Note that the source file also contains the functions fb_number() and fb_vars() besides the function fb_unit(). fb_unit() is the most mature one.

**Comment by sangwinc** [comment]
Thanks so much for posting this code.  This is, I think, the first time someone has posted a library like this and asked us to include it.  So, apologies if we don't yet have a smooth path for this kind of feature inclusion.

I think the functionality is really valuable (clearly, you are using it widely!) and this is something we should consider in the STACK core.  I think one way to proceed would be to add in your functions, and then re-factor the existing answer test code to make use of the same functions as far as is practicable.  Indeed, some of the existing feedback in answer tests might profit from being exposed as separate functions using the fb_... naming convention you use.

Another option is to formalize the libraries, and to keep your work as a self-contained library.  Perhaps we could include it in the STACK release as a separate file?  One goal is to try to make it easier to let people contribute code, and this kind of library is one way to do that.  However, we have a tension with keeping things coherent and in one place, including documentation and unit tests!

One wrinkle: your code includes language strings (of course, in the feedback) and we need to internationalize these.  This is a real barrier here, since we need to publish the strings in the Moodle language packs and then wait for translators to catch up!     Currently this is done with the StackAddFeedback function.

https://github.com/maths/moodle-qtype_stack/blob/master/stack/maxima/stackmaxima.mac#L893

The convention is that the answer note should be the language tag, and I'd like to keep this convention.  Mostly, the notes start with the name of the answer test, e.g.  ATMatrix_wrongentries is part of the Matrix answer test:

https://github.com/maths/moodle-qtype_stack/blob/master/stack/maxima/stackmaxima.mac#L1676
and the language tag is here
https://github.com/maths/moodle-qtype_stack/blob/master/lang/en/qtype_stack.php#L892

The translation is a real barrier to making it easy for people to contribute, and I'm sorry about this.

Do you have any unit tests for your code?
Chris

**Comment by aharjula** [comment]
Chris, regardless of whether Martin actually wants to include portions of his library into STACK core for release, we really should work on the localisation of all things going on in Maxima. From the point of view of both STACK itself and external libraries like MecLib. To me, it seems like the tools we currently have are as follows:

 1. The old way of using specific strings to trigger [string replacement](https://github.com/maths/moodle-qtype_stack/blob/913dc5cd204dbf81b8af14d94c75cd841176415e/locallib.php#L130-L165) which requires that we basically check the whole output coming as feedback for those specific markers.
 2. The `[[commonstring/]]`-block, which directly accesses those language-pack strings and can do injections into named parameters in them.
 3. The `[[lang]]`-block allows one to define some language versions within the code, but those are then hard coded. This will always be more of a question authors tool not a tool for library builders.
 4. The new `[[template]]`-block, which allows complete replacement of whole named sections of CASText, which might not seem like a tool for localisation but is indeed also a tool for it.

Personally, I dislike the old way. It requires all sorts of special handling and is hard to parse. No library should ever use it and I would really like to excise the whole thing from the core as well. What I would like people using would be inline `castext` and those blocks, and the core should use the `[[commonstring/]]` block for every bit of static text (with or without parameters). Ideally, all such texts should be wrapped in named sections with `[[template]]`-blocks so that one could override both the language and actual content at the question level.

From the point of view of external libraries, the situation is more difficult, as for them updating `lang/en/qtype_stack.php` is not really an option, and they thus cannot really use the `[[commonstring/]]`-block for their own strings. It could be that for them we should have the option for retrieving their own strings (probably as JSON) from elsewhere (e.g. `[[commonstring libraryurl="https://.../langs/"/]]`), but that might have some security implications (not much worse than the inclusions themselves). But as long as `[[commonstring/]]` is not the option for them they may limit themselves to specific languages and use the `[[lang]]` and `[[template]]` blocks to deal with things. Note that combining these two provides that very same "libraryurl" feature by providing separate template overrides for different languages and including those when needed:

```
stack_include("...mylibrary/template-overrides-for-EN.mac");
```
This will probably be quite expensive in the sense of unused code being present in questions, but for smaller libraries with small numbers of strings, it should not be too bad.

In the end, as long as we have external libraries that live faster than STACK releases, they too need to have tools for dealing with localisation and we really want to have similar tools in play in the core as well to allow people to use the same methods for both. The solution must never be to freeze a library to the core, unless that library has actually stopped developing on its own, at which point it would be nice if the library already used the same methods and could thus be easily transitioned over.

**Comment by mkraska** [comment]
I tried to use a STACK question das unit test. See te attached question in my earlier post. It didn't really work because I found no way to automatically  compare the feedback output cas text with reference values. Therefore the unit test relies on visual comparison.

**Comment by mkraska** [comment]
Matti, thanks for explaining the options for localization of external includes or libraries. I think that the issue goes beyond just switching languages. Should there be some day any other meclib users exept myself, they certainly will have their own preferences for feedback texts, even if they are german. 

Thus, a custom replacement mechanism might be required and probably must be an external link, because otherwise every question would need to be touched in case of core upgrades or if the trainer changes his mind about appropriate texts.

As a convenience feature, I'd appreciate a way to specify custom replacements on course level, under control of the course admin (trainer).

**Comment by sangwinc** [comment]
We know have a mechanism for this via stack_include()
